src/structure_building_blocks.zig

```diff
@@ -0,0 +1,309 @@
+				return error.NoOriginBlock;
+			}
+			return info;
+		}
+	};
+};
+
+pub fn isChildBlock(block: Block) bool {
+	return childBlockNumericIdMap.contains(block.typ);
+}
+
+pub fn isOriginBlock(block: Block) bool {
+	return block.typ == originBlockNumericId;
+}
+
+const originBlockStringId = "cubyz:sbb/origin";
+var originBlockNumericId: u16 = 0;
+
+// Maps global child block numeric ID to index used to locally represent that child block.
+var childBlockNumericIdMap: std.AutoHashMapUnmanaged(u16, u16) = .{};
+var childBlockStringId: [20][]const u8 = undefined;
+
+pub const StructureBuildingBlock = struct {
+	children: Children,
+	blueprint: [4]*BlueprintEntry,
+
+	fn initFromZon(stringId: []const u8, zon: ZonElement) !StructureBuildingBlock {
+		const blueprintId = zon.get(?[]const u8, "blueprint", null);
+		if(blueprintId == null) {
+			std.log.err("[{s}] Missing blueprint field.", .{stringId});
+			return error.MissingBlueprintIdField;
+		}
+
+		const blueprintRef = blueprintCache.?.get(blueprintId.?);
+		if(blueprintRef == null) {
+			std.log.err("[{s}] Could not find blueprint '{s}'.", .{stringId, blueprintId.?});
+			return error.MissingBlueprint;
+		}
+
+		return StructureBuildingBlock{
+			.children = Children.initFromZon(stringId, zon.getChild("children")),
+			.blueprint = blueprintRef.?,
+		};
+	}
+	fn finalize(self: *StructureBuildingBlock) !void {
+		try self.children.finalize();
+	}
+	pub fn getBlueprint(self: StructureBuildingBlock, rotation: Degrees) *BlueprintEntry {
+		return self.blueprint[@intFromEnum(rotation)];
+	}
+};
+
+const Children = struct {
+	colors: [20]AliasTable(Child),
+
+	fn initFromZon(stringId: []const u8, zon: ZonElement) Children {
+		var self: @This() = .{.colors = undefined};
+		for(childBlockStringId, 0..) |colorName, i| {
+			self.colors[i] = initChildTableFromZon(colorName, stringId, zon.getChild(colorName));
+		}
+		return self;
+	}
+	fn finalize(self: *Children) !void {
+		for(self.colors) |color| {
+			for(color.items) |*c| try c.finalize();
+		}
+	}
+	pub fn pickChild(self: Children, block: BlueprintEntry.Info.StructureBlock, seed: *u64) Child {
+		return self.colors[block.index].sample(seed).*;
+	}
+};
+
+fn initChildTableFromZon(childName: []const u8, stringId: []const u8, zon: ZonElement) AliasTable(Child) {
+	if(zon == .null) return .init(arena_allocator, &[0]Child{});
+	if(zon != .array) {
+		std.log.err("[{s}->{s}] Incorrect child data structure, array expected.", .{stringId, childName});
+		return .init(arena_allocator, &[0]Child{});
+	}
+	if(zon.array.items.len == 0) {
+		std.log.warn("[{s}->{s}] Empty children list.", .{stringId, childName});
+		return .init(arena_allocator, &[0]Child{});
+	}
+	var list = arena_allocator.alloc(Child, zon.array.items.len);
+	for(zon.array.items, 0..) |entry, i| {
+		list[i] = Child.initFromZon(childName, stringId, i, entry);
+	}
+	return .init(arena_allocator, list);
+}
+
+const Child = struct {
+	structureId: []const u8,
```

Comment:
Child structures must be resolved in second pass, as they may not be loaded when parent structure is, hence ID is required to fetch correct structure then.

Reply:
Well then please store a separate data structure for the unresolved entries that's cleared after loading.
There is no need for this data to persist after loading. It's just clogging the cache lines.

Reply:
You could literally just store a **StructureBuildingBlock to this location and it's string id in a list, then resolve them easily.

Reply:
Do they have guaranteed pointer stability? There is return value optimization, but is that actually guaranteed?
And this will also mean that `structureCache` has to be `std.StringHashMapUnmanaged(*StructureBuildingBlock)`, not `std.StringHashMapUnmanaged(StructureBuildingBlock)`, unless entries have guaranteed pointer stability after inserting new elements?

Reply:
Ah right, I guess you have to use id and child index instead then.